cdi/src/main/java/io/smallrye/context/inject/SmallryeContextCdiExtension.java

```diff
@@ -371,4 +372,44 @@ public void cleanup(@Observes AfterDeploymentValidation adv) {
         this.executorMap.clear();
         this.threadContextMap.clear();
     }
+
+    // Used to delay the ConfigProvider.getConfig() call until we need the config
+    // Doing it too early may break application deployment in application servers
+    private static class LazyConfig implements Config {
+        private volatile Config delegate;
```

Comment:
@kabir Why do you use `volatile` (and the `sync` block)?
I don't think we are going to access CDI extension in parallel, plus the code here that uses `Config` is executed once during bootstrap and that's it.

Reply:
Basically to make sure that it is safely published if different threads use it. The cost of this is minimal. I need to dig deeper to validate what you are saying if I am to remove it :)

Reply:
I don't see how this can be used in different threads, maybe I am missing something?
It's a CDI extension and as such it is a part of CDI bootstrap process which should invoke all extensions in one thread.

Reply:
@manovotn I'm not saying it IS happening in different threads, more that I am not familiar in depth with CDI, so I need to check that it really is all happening in one thread

Reply:
Sorry, should have given more background. Basically logic within CDI extensions is based on observer notification and all of those are all synchronous observers (blocking, on the main thread).
Furthermore the method notification for lifecycle observers is already [synchronized in Weld](https://github.com/weld/core/blob/master/impl/src/main/java/org/jboss/weld/event/ExtensionObserverMethodImpl.java#L122).

Therefore, the only case where I can see this failing is if you temper with it and try to invoke stuff manually, which you shouldn't in the first place ;)

Reply:
@manovotn My original attempt was to not use this wrapper class and to just initialise it in the BeforeBeanDiscovery method. This is simpler and I assume this is guaranteed to happen before the other ones. Do you prefer that?

Reply:
Helper class is also fine, whatever you like.
All I am asking is to get rid of the synchronization as I am pretty convinced it's superfluous.